nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/ConnectionLoadBalanceServer.java

```diff
@@ -0,0 +1,251 @@
+            threadPool.submit(action);
+        }
+
+        this.communicationActions = actions;
+
+        acceptConnection = new AcceptConnection(serverSocket);
+        final Thread receiveConnectionThread = new Thread(acceptConnection);
+        receiveConnectionThread.setName("Receive Queue Load-Balancing Connections");
+        receiveConnectionThread.start();
+    }
+
+    public int getPort() {
+        return serverSocket.getLocalPort();
+    }
+
+    public void stop() {
+        stopped = false;
+        threadPool.shutdown();
+
+        if (acceptConnection != null) {
+            acceptConnection.stop();
+        }
+
+        communicationActions.forEach(CommunicateAction::stop);
+
+        Socket socket;
+        while ((socket = connectionQueue.poll()) != null) {
+            try {
+                socket.close();
+                logger.info("{} Closed connection to {} on Server stop", this, socket.getRemoteSocketAddress());
+            } catch (final IOException ioe) {
+                logger.warn("Failed to properly close socket to " + socket.getRemoteSocketAddress(), ioe);
+            }
+        }
+    }
+
+    private ServerSocket createServerSocket() throws IOException {
+        final InetAddress inetAddress = hostname == null ? null : InetAddress.getByName(hostname);
+
+        if (sslContext == null) {
+            return new ServerSocket(port, 50, InetAddress.getByName(hostname));
+        } else {
+            final ServerSocket serverSocket = sslContext.getServerSocketFactory().createServerSocket(port, 50, inetAddress);
+            ((SSLServerSocket) serverSocket).setNeedClientAuth(true);
+            return serverSocket;
+        }
+    }
+
+
+    private class CommunicateAction implements Runnable {
+        private final LoadBalanceProtocol loadBalanceProtocol;
+        private volatile boolean stopped = false;
+
+        public CommunicateAction(final LoadBalanceProtocol loadBalanceProtocol) {
+            this.loadBalanceProtocol = loadBalanceProtocol;
+        }
+
+        public void stop() {
+            this.stopped = true;
+        }
+
+        @Override
+        public void run() {
+            String peerDescription = "<Unknown Client>";
+
+            while (!stopped) {
+                Socket socket = null;
+                try {
+                    socket = connectionQueue.poll(1, TimeUnit.SECONDS);
+                    if (socket == null) {
+                        continue;
+                    }
+
+                    peerDescription = socket.getRemoteSocketAddress().toString();
+
+                    if (socket.isClosed()) {
+                        logger.debug("Connection to Peer {} is closed. Will not attempt to communicate over this Socket.", peerDescription);
+                        continue;
+                    }
+
+                    logger.debug("Receiving FlowFiles from Peer {}", peerDescription);
+                    loadBalanceProtocol.receiveFlowFiles(socket);
```

Comment:
Receiving side should implement some way to report that the node cannot receive any more FlowFiles because of back-pressure.

![image](https://user-images.githubusercontent.com/1107620/46066713-14abac00-c1b0-11e8-8b3b-2db86740aa9f.png)


The sending side may keep sending more FlowFiles even the number of queued FlowFiles exceeds the configured back-pressure. Scenario, two-node cluster, nifi0 and nifi1. NiFi0 is the primary, and a connection is configured to use Partition By Attribute. The primary node generates FlowFiles having attribute so that they will be sent to nifi1. NiFi0 only consider the number of queued FlowFile locally to schedule the upstream processor.

Reply:
That's a good point, and honestly not something that I had thought about. Whether or not we would need to test for this would depend on the select partitioner. For example, if trying to offload all data from the node for decommissioning purposes, we want to ignore that, I think, and push data to the node anyway. In the case of Partition By Attribute, though, you have a great point.
Personally, I would be inclined to create another JIRA to address this in the future, and document this as a limitation for the time being. This will take a decent bit of thought to do this efficiently and effectively.

Reply:
In addition to documentation, I do think it would be nice in the UI, when configuring Load Balancing, to indicate that Backpressure Thresholds may not be honored when using load balancing. I also created NIFI-5651 to track this.

Reply:
@ijokarumawak please ignore my comment above :) After thinking about it a good bit more, I realized that the way I was thinking about it previously was far more complicated than it needed to be. I have pushed one more commit that will take backpressure into account.

It works by first checking with the remote node to see if there is space in the queue or if backpressure is applied. If it's applied, the node won't send data. This will cause the data to queue up on the 'client' node, which will cause backpressure to engage there as well.